server/server.go

```diff
@@ -1540,3 +1554,98 @@ func (s *Server) serviceListeners() []net.Listener {
 	}
 	return listeners
 }
+
+// Returns true if in lame duck mode.
+func (s *Server) isLameDuckMode() bool {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.ldm
+}
+
+// This function will close client and route listeners
+// then close clients at some interval to avoid a reconnecting storm.
+// Note that the routes are closed before so that when a client is
+// disconnected and reconnects to another server, that server does
+// not send to this server a route SUB.
+func (s *Server) lameDuckMode() {
+	s.mu.Lock()
+	// If we are in the process (or done) or if listener is not yet
+	// ready, just return.
+	if s.ldm || s.listener == nil {
+		s.mu.Unlock()
+		return
+	}
+	expected := 1
+	s.Noticef("Entering lame duck mode, stop accepting new clients")
+	s.ldm = true
+	s.ldmCh = make(chan bool, 1)
+	s.ldmDoneCh = make(chan bool, 1)
+	s.listener.Close()
+	s.listener = nil
+	if s.routeListener != nil {
+		s.routeListener.Close()
+		s.routeListener = nil
+		expected++
+	}
+	s.mu.Unlock()
+
+	// Wait for accept loop to be done to make sure that no new
+	// client can connect
+	for i := 0; i < expected; i++ {
+		<-s.ldmCh
+	}
+
+	durInMs := int64(s.getOpts().LameDuckDuration / time.Millisecond)
+	if durInMs < 1 {
+		durInMs = 1
+	}
+
+	// Close routes
+	s.mu.Lock()
+	routes := make([]*client, 0, len(s.routes))
+	for _, r := range s.routes {
+		routes = append(routes, r)
+	}
+	s.mu.Unlock()
+	for _, r := range routes {
+		r.setRouteNoReconnectOnClose()
+		r.closeConnection(LameDuckMode)
+	}
+
+	// Now capture all clients
+	s.mu.Lock()
+	numClients := int64(len(s.clients))
```

Comment:
I would try to simplify. I would fire a timer every second and pre calculate batch size of how many to close per timer fire. Once zero your done.

Reply:
Will think about it. Thanks!

Reply:
Sounds good, thanks for doing this. I think this will be a huge help to NGS.

Reply:
The thing is that for test, I don't want to have it take so much time, this is why I was doing it this way where code still "works" even with very low duration (in the ms range). Still, will think about it more.

Reply:
You could configure the timer interval for sure.

Reply:
I have some pending changes, but still did not implement that. The more I think about it, how is that better? Say there are 1000 clients to close within 30 seconds. There is timer firing every 1 sec, what do you expect the timer callback to do? Close a certain number of clients in a row then reset the timer to fire in 1 sec? I am not sure that this is good because there will still be a bunch of reconnect at once, as opposed to spread the closing of the clients (still may require to close more than one in a row if there is many more clients that the number of ms left).

Reply:
That is, the pause of 1sec between each interval is helping I guess, but not as much as making sure that - if we can - we spread the closing, I think.

Reply:
I think you are right. Even though we expect clusters of a size that all of the batch clients connect to not the same machine, spreading it out a bit even within a batch makes more sense. We should pick a timer interval that allows us 1 client close per tick (as best we can).

Reply:
Honestly, using a timer here makes the code way more complex than the original in the PR (although I would change the 10ms sleep between batches. Are you sure about the timer approach?